feat: add XvB options, refactor & improve algorithm

mostafaei2002 opened the pull request:
- Add Simple/Advanced panels to XVB Tab
- Manually set donation hashrate for xvb or p2pool

Cyrix126 commented:
# Changes needed to merge
## UI:
### Same UI for simple sub-menu
the UI for simple tab should not be different from before. The hero checkbox should not be replaced by a drop down menu.
The reason is that a checkbox when you have only have two choice makes more sense since you need to make only one click instead of two to choose the other mode. Since Gupaxx is intended to be a easy to use GUI, this design detail is important.
### keep XvB logo
The logo of XvB should not be removed as it is a link to the website easily seen. If there is not enough height space, you must  add scroll bar.
### Optional: Slider instead of input
Instead of having an input, a slider with % of HR or a number based on runtime xmrig would prevent user putting numbers above what their rig can offer.
## Backend:
### Pass all the CI checks
You can enable them for your branch by setting main to dev in .github/workflows/*.yml

mostafaei2002 commented:
### Requested Changes
- Reverted the Simple Tab UI to the previous state
- fixed XvB logo & added a scrollbar to fix height issues
-  Added a slider for advanced tab

### New feature
- Added a new mode named ManualDonationLevel which allows users to manually select the donation tier e.g Donor, DonorVIP, etc they want to participate in.
- fixed failing tests

Cyrix126 commented:
# Changes needed to merge

## UI
- Make sliders with a way to choose unit scale like in the sub-menu p2pool on Status Tab.
- hovering message for manually donate and manually keep  should be specific to the mode chosen.
- ManualDonationLevel mode should have hovering message on each round type with the minimum HR donated needed.

## Backend
- add tests to check if each mode will give the right amount of time

mostafaei2002 commented:
All requested changes were implemented.

### Ideas
- Instead of disabling the slider when xvb is not alive change it to when xmrig hashrate is lower than a number. IMO users should be able to set the donation amount before xvb getting started. This improves UX.
- Change the names for ManuallyDonate, ManuallyKeep names. I am thinking of ManualXvb instead of ManuallyDonate  and ManualP2pool instead of ManuallyKeep. These names are more self explanatory.

Cyrix126 commented:
>     * Instead of disabling the slider when xvb is not alive change it to when xmrig hashrate is lower than a number. IMO users should be able to set the donation amount before xvb getting started. This improves UX.

I agree. But let's say it is limited on HR numbers from xmrig after the first 10 seconds are obtained, like the algo do. 

And adding a hovering help message so users understand that their set value can change if it's too much for xmrig.

For maximum value before the 10s you can set a thousand so that the slider will scale nicely (it's the same in Status P2Pool sub). 
MH/s should be also added in this case. With the decimal separator, users will have good control over the value.
 
The numbers after decimal  should be no more than 3 to be like the account stats.

>     * Change the names for ManuallyDonate, ManuallyKeep names. I am thinking of ManualXvb instead of ManuallyDonate  and ManualP2pool instead of ManuallyKeep. These names are more self explanatory.

ok good.

mostafaei2002 commented:
All the changes are implemented plus a few extra minor changes. There could be some other changes here and there that are needed I'll proceed to run the program and test it in the upcoming days.

Cyrix126 commented:
rust formatting checks failed. Since your branch is not up to date with upstream, your CI runs were missing the new ones.

Cyrix126 commented on src/disk/state.rs:
```diff
@@ -243,10 +243,43 @@ pub struct Xmrig {
 }
 
-#[derive(Clone, Eq, PartialEq, Debug, Deserialize, Serialize, Default)]
+#[derive(Clone, PartialEq, Debug, Deserialize, Serialize, Default)]
```
Rust default of a bool is false but simple should be true the first time the value is created in the state. Else the XvB tab will open to advanced tab at next update. Simple tab should always be preferred as default.

Cyrix126 commented on src/app/panels/middle/xvb.rs:
```diff
@@ -23,235 +30,362 @@ use crate::{
+                                    lock!(gui_api_xmrig).hashrate_raw_1m
+                                } else if lock!(gui_api_xmrig).hashrate_raw > 0.0 {
+                                    lock!(gui_api_xmrig).hashrate_raw
+                                } else {
```
slider should still be enabled here.

mostafaei2002 replied:
I am not sure about this one. Why should the slider be enabled before xmrig is started. Right now the slider will get enabled as soon as xmrig reports any hashrate above 0. Why should the slider be enabled before this? Before this point in time the slider shows 1000 hashrate which is not the real hashrate. I don't see any point in the slider being enabled before xmrig reports the hashrate.

Cyrix126 replied:
> I don't see any point in the slider being enabled before xmrig reports the hashrate.

We want the user to be able to set it without anything started yet, as he can know in advance how much HR he will get and doesn't want to wait the start of xmrig. They would just have to launch Gupaxx with everything auto starting.

 I thought that's what we talked about [before](https://github.com/Cyrix126/gupaxx/pull/11#issuecomment-2161117694) but I think I didn't state clearly it was not only before xvb but also before xmrig. If the value set is too much when real HR appear, it will be overwritten for runtime but not saved.

> And adding a hovering help message so users understand that their set value can change if it's too much for xmrig.

That's why.


> Touching the slider before the hashrate starts coming in also overrides the configuration from previous sessions.

It should not be saved unless "Save" button is clicked on but should be applied to next algo anyway (same as current hero mode behavior).

Also I just realized that the HR values do not get cleaned when xmrig is stopped. I fixed it in https://github.com/Cyrix126/gupaxx/commit/5c1266cba2bb94fb94c7602ef5f19205400ebf12.

mostafaei2002 replied:
Sorry, I had misunderstood your point previously. That makes sense.

Cyrix126 commented on src/app/panels/middle/xvb.rs:
```diff
@@ -23,235 +30,362 @@ use crate::{
+                        egui::ComboBox::from_label("")
+                        .selected_text(format!("{:?}", self.mode))
+                        .show_ui(ui, |ui| {
+                                ui.selectable_value(&mut self.mode, XvbMode::Auto, "Auto");
+                                ui.selectable_value(&mut self.mode, XvbMode::Hero, "Hero");
+                                ui.selectable_value(&mut self.mode, XvbMode::ManualXvb, "ManualXvb")
+                                .on_hover_text(XVB_MODE_MANUAL_XVB_HELP);
+                                ui.selectable_value(&mut self.mode, XvbMode::ManualP2pool, "ManualP2pool")
+                                .on_hover_text(XVB_MODE_MANUAL_P2POOL_HELP);
+                                ui.selectable_value(&mut self.mode, XvbMode::ManualDonationLevel, "ManualDonationLevel")
+                                .on_hover_text(XVB_MODE_MANUAL_DONATION_LEVEL_HELP);
```
drop menu choices and selected choice should both have space in the name.

Cyrix126 commented on src/app/panels/middle/xvb.rs:
```diff
@@ -151,17 +151,26 @@ impl crate::disk::state::Xvb {
+                                ui.selectable_value(&mut self.mode, XvbMode::ManualP2pool, "Manual P2pool")
                                 .on_hover_text(XVB_MODE_MANUAL_P2POOL_HELP);
-                                ui.selectable_value(&mut self.mode, XvbMode::ManualDonationLevel, "ManualDonationLevel")
+                                ui.selectable_value(&mut self.mode, XvbMode::ManualDonationLevel, "Manual Donation Level")
```
replace this modification  by implementing Display trait on XvBMode enum and calling directly self.mode.to_string()

Cyrix126 commented on src/app/panels/middle/xvb.rs:
```diff
@@ -222,11 +231,11 @@ impl crate::disk::state::Xvb {
+                            ui.radio_value(&mut self.manual_donation_level, ManualDonationLevel::DonorWhale, "Donor Whale")
                             .on_hover_text(XVB_DONATION_LEVEL_WHALE_DONOR_HELP);
-                            ui.radio_value(&mut self.manual_donation_level, ManualDonationLevel::DonorMega, "DonorMega")
+                            ui.radio_value(&mut self.manual_donation_level, ManualDonationLevel::DonorMega, "Donor Mega")
```
implement Display trait on ManualDonationLevel enum and calling directly self.manual_donation_level.to_string()

mostafaei2002 replied:
I don't see how I would do that since instead I used a Display trait and for each option did the below
```
ManualDonationLevel::Donor.to_string()
```

Cyrix126 replied:
I missread, thought there was two different enums. It looks good with your last commit.

Cyrix126 replied:
that's what I meant :)

Cyrix126 commented on src/disk/state.rs:
```diff
@@ -399,6 +399,21 @@ impl Default for Xmrig {
     }
 }
 
+impl Default for Xvb {
+    fn default() -> Self {
+        Self {
+            simple: true,
+            token: "".to_string(),
+            simple_hero_mode: false,
+            mode: XvbMode::Auto,
+            manual_amount_raw: 0.0,
+            manual_slider_amount: 0.0,
+            manual_donation_level: ManualDonationLevel::Donor,
+            manual_donation_metric: ManualDonationMetric::Hash,
+        }
+    }
+}
+
```
use ..default::Default() instead of writing every fields for the ones that still use the default value.

Cyrix126 replied:
did you forgot it ? you marked it as resolved.

Cyrix126 reviewed (changes requested):
need a hover help for slider to precise that too much Hr can be overwritten.
and rewrite impl Default for XvB with shorter code by using ..default::Default()

Cyrix126 commented:
need to fix:
- errors with xmrig advanced args
- estimation more than 600 overflow
- real HR lower than specified.

Cyrix126 commented on src/helper/p2pool.rs:
```diff
@@ -491,13 +491,15 @@ impl Helper {
 
         // 4. Loop as watchdog
         let mut first_loop = true;
+        let mut last_p2pool_request = Arc::new(Mutex::new(tokio::time::Instant::now()));
+        let mut last_status_request = Arc::new(Mutex::new(tokio::time::Instant::now()));
```
any reason to use Arc<Mutex<_>> ? I think we can skip the overhead.

mostafaei2002 replied:
No specific reason, I looked into the other files and this is how it was implemented so I went along with it.

Cyrix126 replied:
You can remove the Arc<Mutex<_>> . It is wrapped into it in [spawn_xvb_watchdog](https://github.com/Cyrix126/gupaxx/blob/d706dec3fcb861f34dae0ebbd9f1a995656b9ec8/src/helper/xvb/mod.rs#L151) method since it is shared accross a tokio spawn, which is not the case here.
